refactor: [M3-9354] - Reroute Firewalls Feature

abailly-akamai opened the pull request:
## Description 📝
Migrating our firewalls feature to Tanstack router

<img src="https://github.com/user-attachments/assets/dafc00ae-5609-4294-aecc-8cc3d66180b4" width="300" />


> [!NOTE]
> Despite the high count of files in this PR, this is a fairly light refactor and I was careful not to go overboard with improving what is a bit of an old feature (in the way it was composed/state managed) 

## Changes  🔄
 **Router Migration**
- Removed the old React Router implementation (`react-router-dom`)
- Added new TanStack Router routes and configurations
- Created new lazy routes for Firewalls feature

 **Component Updates**
- Updated components to use TanStack Router hooks (`useParams`, `useNavigate`, etc.)
- Refactored navigation logic to use the new router
- Updated test files to mock TanStack Router hooks
- Better handlers naming conventions

**Route Structure Changes**
- Added more granular routes for different firewall actions
- Implemented nested routes for features like rules, linodes, and nodebalancers
- Routed modals and drawers
- Added proper route validation and parameter handling

**Code Organization**
- Moved constants to a separate file (`constants.ts`)
- Consolidated route definitions in dedicated files
- Improved type safety and parameter handling

**Components Tests Setup Update
- Allow components routed/sorted/paginated via tanstack to run

## Preview 📷
There should be no visual change or regression as a part of this PR

## How to test 🧪

### Verification steps
- Confirm Firewall Flows (`/firewalls`)
  - `/firewalls` > create, disable, delete
  - /firewalls/{id} > rules, linodes and nodebalancers actions
- Confirm Firewall Create Flow elsewhere
  - `/linodes/create` > create firewall
  - /linodes/63686452/networking` > add, unassign
- Unit tests pass
- CY e2e tests pass
- CY component tests pass

<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

👀 Doing a self review
❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
🤏 Splitting feature into small PRs
➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
🧪 Providing/improving test coverage
 🔐 Removing all sensitive information from the code and PR description
🚩 Using a feature flag to protect the release
👣 Providing comprehensive reproduction steps
📑 Providing or updating our documentation
🕛 Scheduling a pair reviewing session
📱 Providing mobile support
♿  Providing accessibility support
</details>

- [x] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [x] All unit tests are passing
- [x] TypeScript compilation succeeded without errors
- [x] Code passes all linting rules

abailly-akamai commented on packages/manager/cypress/support/component/setup.tsx:
```diff
@@ -38,10 +47,23 @@ import type { ThemeName } from '@linode/ui';
+      initialEntries: ['/'],
+    }),
+    routeTree: rootRoute.addChildren([indexRoute]),
+  });
```
CY component tests did not have support for Tanstack router tooling yet

abailly-akamai commented on packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceTable.tsx:
```diff
@@ -1,43 +1,46 @@
+    );
+    const devices =
+      allDevices?.filter((device) => device.entity.type === type) || [];
+
```
Moved the query closer to the component which helps with sorting/pagination tools

abailly-akamai commented on packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceTable.tsx:
```diff
@@ -47,67 +50,77 @@ export const FirewallDeviceTable = React.memo(
+      orderBy,
+      sortedData: sortedDevices,
+    } = useOrderV2({
+      data: devices,
```
reminder - using `data` means client side sorting. 

I did try to sort by label via Z-Filter and although it appears to be supported it did not work for me.

This was the behavior previously so am not overly concerned here - I'll inquire about this with API team however

abailly-akamai commented on packages/manager/src/features/Firewalls/FirewallDetail/Devices/constants.ts:
```diff
@@ -0,0 +1,7 @@
+import type { FirewallDeviceEntityType } from '@linode/api-v4';
+
+export const formattedTypes: Record<FirewallDeviceEntityType, string> = {
+  interface: 'Interface', // @TODO Linode Interface: double check this when working on UI tickets
```
@coliu-akamai heads up this moved to the `constants.ts` file 👁️

mjac0bs reviewed (commented):
This looks really close!

Main Firewall flows:
✅  /firewalls > create, disable, delete
⚠️ `/firewalls/{id}` > rules
✅  `/firewalls/{id}` > linodes actions, including table sorting
✅  `/firewalls/{id}` > nodebalancers actions, including table sorting

Confirm Firewall Create Flow elsewhere
✅ `/linodes/create` > create firewall
✅ `/linodes/63686452/networking` > add, unassign

✅  Unit tests pass
✅  CY e2e tests pass
✅  CY component tests pass

Where I ran into a bit of trouble was the firewall rules editing and the back button. What do we want to do about the "Discard Changes" modal? It should only be displaying when the user tries to navigate somewhere other than `/rules`.

Observations:
- If a rule has been added or edited (but not saved) and the back button is hit, there is either a flash of drawer content or sometimes(?) the drawer opens fully *and* the `Discard Firewall changes?` modal displayed.

https://github.com/user-attachments/assets/b53ffcfd-7794-4e14-92f1-0d481e02e22d

- If a rule has been cloned or deleted (but not saved) and the back button is hit to go back to the landing page, no `Discard Firewall changes?` modal is displayed, which is a regression from prod. If other actions have already been taken with the rule, we'll route back to the drawer and display the modal.

| Prod | This Branch |
| -------- | -------- |
| <video src="https://github.com/user-attachments/assets/8343b22f-cce5-447b-b8e9-695b7109612e"/> | <video src="https://github.com/user-attachments/assets/d7527143-6b5c-48d2-9ae6-9643017a3684"/> | 

- If a default outbound or inbound policy has been changed (but not saved) and the back button is hit, no `Discard Firewall changes?` modal is displayed, which is a regression from prod.

| Prod | This Branch |
| --------| -------- |
| <video src="https://github.com/user-attachments/assets/c2656bb8-e7f6-4057-9bf6-49134317ce34"/> | <video src="https://github.com/user-attachments/assets/4838bf13-2162-41e2-acc5-fcddc45c29b1"/> |

coliu-akamai commented:
Heads up - just merged in #11678,  which looks like it caused a conflict on CreateFirewallDrawer.tsx (looking at your changes for that file, it should be super small though 🤞)

abailly-akamai commented:
I have a decent path forward to replicate this behavior with the new router, but this will require to update tanstack router to a later version, which is outside the scope here. putting this back into draft before addressing feedback and conflicts. Yay!

abailly-akamai commented:
@mjac0bs I did (not without a fair amount of self inflicted pain) managed to preserve the unsaved rules prompt logic with this migration. Please test again at your convenience and confirm the bheavior. Thank you!

mjac0bs commented on packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceActionMenu.tsx:
I found one edge-case issue related to backwards/forwards route navigation and the remove device dialog.

`device` will be`undefined` when returning to a previously visited `/<device type>/remove` route. Without a valid `device`, we always default to the NodeBalancer as the device type in the remove dialog and there is no device entity label ([code](https://github.com/linode/manager/blob/develop/packages/manager/src/features/Firewalls/FirewallDetail/Devices/RemoveDeviceDialog.tsx#L36-L47)). We can't successfully complete the action; so should we [return](https://github.com/linode/manager/blob/develop/packages/manager/src/features/Firewalls/FirewallDetail/Devices/RemoveDeviceDialog.tsx#L90-L91) (`device && <ConfirmationDialog...`)?

https://github.com/user-attachments/assets/67d1b68b-3b40-4a61-993b-bc2389082005

mjac0bs reviewed (approved):
I think this is looking good now from my read-through and testing, pending one minor issue I noticed with the remove device dialog.

I carefully QAed the discard changes flows and it looks like we've got parity with prod now with edits to the firewall rules, thanks! 🎉 

https://github.com/user-attachments/assets/b902c19e-bfe5-4145-a6bd-5c857da6d6a6

harsh-akamai reviewed (approved):
✅  Confirmed Firewall Flows 
✅  Confirm Firewall Create Flow elsewhere
✅  Unit tests pass
✅  CY e2e tests pass
✅  CY component tests pass

abailly-akamai commented:
@mjac0bs yup! good catch - addresssed in https://github.com/linode/manager/pull/11704/commits/4ca2cf1ddc27cce5c0db46ed4582a61c37954b0a
